tests/decomposition/test_decomposition_graph.py

```diff
@@ -0,0 +1,181 @@
+        def custom_hadamard(wires):
+            qml.PhaseShift(np.pi / 2, wires=wires)
+            qml.RX(np.pi / 2, wires=wires)
+            qml.PhaseShift(np.pi / 2, wires=wires)
+
+        @qml.register_resources({qml.PhaseShift: 1, qml.RY: 1})
+        def custom_hadamard_2(wires):
+            qml.PhaseShift(np.pi / 2, wires=wires)
+            qml.RY(np.pi / 2, wires=wires)
+
+        graph = DecompositionGraph(operations=[qml.Hadamard(0)], target_gate_set={"RX", "RY", "RZ"})
+        assert graph._get_decompositions(qml.Hadamard) == decompositions[qml.Hadamard]
+
+        graph = DecompositionGraph(
+            operations=[qml.Hadamard(0)],
+            target_gate_set={"RX", "RY", "RZ"},
+            fixed_decomps={qml.Hadamard: custom_hadamard},
+        )
+        assert graph._get_decompositions(qml.Hadamard) == [custom_hadamard]
+
+        graph = DecompositionGraph(
+            operations=[qml.Hadamard(0)],
+            target_gate_set={"RX", "RY", "RZ"},
+            alt_decomps={qml.Hadamard: [custom_hadamard, custom_hadamard_2]},
+        )
+        assert (
+            graph._get_decompositions(qml.Hadamard)
+            == [
+                custom_hadamard,
+                custom_hadamard_2,
+            ]
+            + decompositions[qml.Hadamard]
+        )
+
+    @pytest.mark.unit
+    def test_graph_construction(self, _):
+        """Tests constructing a graph from a single Hadamard."""
+
+        op = qml.Hadamard(wires=[0])
+        graph = DecompositionGraph(operations=[op], target_gate_set={"RX", "RZ", "GlobalPhase"})
+        # 5 ops and 3 decompositions (2 for Hadamard and 1 for RY)
+        assert len(graph._graph.nodes()) == 8
+        # 8 edges from ops to decompositions and 3 from decompositions to ops
+        assert len(graph._graph.edges()) == 11
+
+        # Check that graph construction stops at gates in the target gate set.
+        graph2 = DecompositionGraph(operations=[op], target_gate_set={"RY", "RZ", "GlobalPhase"})
+        # 5 ops and 2 decompositions (RY is in the target gate set now)
+        assert len(graph2._graph.nodes()) == 7
+        # 6 edges from ops to decompositions and 2 from decompositions to ops
+        assert len(graph2._graph.edges()) == 8
+
+    @pytest.mark.unit
+    def test_graph_solve(self, _):
+        """Tests solving a simple graph for the optimal decompositions."""
+
+        op = qml.Hadamard(wires=[0])
+        graph = DecompositionGraph(
+            operations=[op],
+            target_gate_set={"RX", "RY", "RZ", "GlobalPhase"},
+        )
+        graph.solve()
+
+        # verify that the better decomposition rule is chosen when both are valid.
+        expected_resource = Resources(
+            {
+                qml.resource_rep(qml.RZ): 1,
+                qml.resource_rep(qml.RY): 1,
+                qml.resource_rep(qml.GlobalPhase): 1,
+            },
+        )
+        assert graph.resource_estimate(op) == expected_resource
```

Comment:
It was not entirely clear to me until this point of reviewing that we get estimates for sub-ops of the originally decomposed ops, and even for ops that only appear during the decomposition :o 
In my opinion, this is a _very_ cool feature, maybe we want it to be mentioned somewhere just to some minimal extent? 
I guess strictly speaking, this is more of internal code and the code + tests tell this story, but it still might be nice to mention :)

Reply:
I added the explanation to the docstring of `DecompositionGraph.resource_estimate`